src/hotspot/share/gc/shenandoah/shenandoahFreeSet.hpp

```diff
@@ -28,63 +30,186 @@
+    assert (which_set > NotFree && which_set < NumFreeSets, "selected free set must be valid");
+    return _capacity_of[which_set];
+  }
+
+  inline size_t used_by(ShenandoahFreeMemoryType which_set) const {
+    assert (which_set > NotFree && which_set < NumFreeSets, "selected free set must be valid");
+    return _used_by[which_set];
+  }
+
+  inline void set_capacity_of(ShenandoahFreeMemoryType which_set, size_t value) {
+    assert (which_set > NotFree && which_set < NumFreeSets, "selected free set must be valid");
+    _capacity_of[which_set] = value;
+  }
 
-  size_t _capacity;
-  size_t _used;
+  inline void set_used_by(ShenandoahFreeMemoryType which_set, size_t value) {
+    assert (which_set > NotFree && which_set < NumFreeSets, "selected free set must be valid");
+    _used_by[which_set] = value;
+  }
 
-  void assert_bounds() const NOT_DEBUG_RETURN;
+  inline size_t max() const { return _max; }
+
+  inline size_t count(ShenandoahFreeMemoryType which_set) const { return _region_counts[which_set]; }
+
+  // Assure leftmost, rightmost, leftmost_empty, and rightmost_empty bounds are valid for all free sets.
+  // Valid bounds honor all of the following (where max is the number of heap regions):
+  //   if the set is empty, leftmost equals max and rightmost equals 0
+  //   Otherwise (the set is not empty):
+  //     0 <= leftmost < max and 0 <= rightmost < max
+  //     the region at leftmost is in the set
+  //     the region at rightmost is in the set
+  //     rightmost >= leftmost
+  //     for every idx that is in the set {
+  //       idx >= leftmost &&
+  //       idx <= rightmost
+  //     }
+  //   if the set has no empty regions, leftmost_empty equals max and rightmost_empty equals 0
+  //   Otherwise (the region has empty regions):
+  //     0 <= lefmost_empty < max and 0 <= rightmost_empty < max
+  //     rightmost_empty >= leftmost_empty
+  //     for every idx that is in the set and is empty {
+  //       idx >= leftmost &&
+  //       idx <= rightmost
+  //     }
+  void assert_bounds() NOT_DEBUG_RETURN;
+};
 
-  bool is_mutator_free(size_t idx) const;
-  bool is_collector_free(size_t idx) const;
+class ShenandoahFreeSet : public CHeapObj<mtGC> {
+private:
+  ShenandoahHeap* const _heap;
+  ShenandoahSetsOfFree _free_sets;
 
   HeapWord* try_allocate_in(ShenandoahHeapRegion* region, ShenandoahAllocRequest& req, bool& in_new_region);
+
+  // While holding the heap lock, allocate memory for a single object which is to be entirely contained
+  // within a single HeapRegion as characterized by req.  The req.size() value is known to be less than or
+  // equal to ShenandoahHeapRegion::humongous_threshold_words().  The caller of allocate_single is responsible
+  // for registering the resulting object and setting the remembered set card values as appropriate.  The
+  // most common case is that we are allocating a PLAB in which case object registering and card dirtying
```

Comment:
Nit pick, but isn't this called for all allocations that are not humongous? Why do we say that PLAB is the most common? Also, much of this comment only applies to the generational mode.

Reply:
Thanks for catching the inappropriate mention of generational issues here.  I'll remove this.

I still think it is the case that the most common allocation is a TLAB and/or GCLAB.  It is only when TLAB/GCLAB allocations fail that we try to do a "shared allocation" of exactly the size of the object to be allocated.

Reply:
Oh yes, I agree with that. Maybe just say `LAB` allocations then? PLABs are specifically for promotions, aren't they?